Add style guide and improve documentation

hillaryj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

Adds a style guide and branch protection guide, plus additional project setup documentation and making sure to apply our own style guide to our docs.

## 💭 Motivation and Context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

These documents will help us work together cohesively. The style guide is a living document and will be added to over time as we find more unwritten rules and guidelines to write.

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Documentation-only - linters pass.

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [x] All new and existing tests pass.

hillaryj committed:
Add branch protection guide and link

hillaryj committed:
Create style guide initial draft

Creates the file for the Style Guide with initial content

hillaryj committed:
Fix header levels

It was too indented and now it is not.

hillaryj committed:
Add content to several sections

hillaryj committed:
Add branch protection and coveralls info

hillaryj committed:
Merge branch 'develop' into add-style-guide

jsf9k commented on style-guide.md:
```diff
@@ -0,0 +1,109 @@
+
+Use sentence case. For example:
+
+- `# Use variables to configure pipelines`
+- `## Use the Style Guide`
```
Historically we have pushed for opening _and closing_ octothorpes.  This is probably worth mentioning in the text for this section, as well as making this change in your examples.
```suggestion
- `# Use variables to configure pipelines #`
- `## Use the Style Guide ##`
```

jsf9k replied:
It appears that sentence case is not actually used for "first-level" (one octothorpe) headings.  This should be documented.

jsf9k commented on style-guide.md:
```diff
@@ -0,0 +1,109 @@
+
+**All repositories should be public.**
+
+Whether creating new repositories or transferring existing repositories,
+remember that we work in the open by default. For more, when we might use
+private repositories, read our [open source policy](https://github.com/cisagov/development-guide/tree/develop/open-source-policy).
```
```suggestion
Whether creating new repositories or transferring existing repositories,
remember that we work in the open by default. For a discussion of the
rare cases in which we might use private repositories, read our 
[open source policy](https://github.com/cisagov/development-guide/tree/develop/open-source-policy).
```

jsf9k commented on project_setup/README.md:
```diff
@@ -89,17 +99,43 @@ Use the following commands to push the new repository to github:
+1. Add a `New repository secret` and name it `COVERALLS_REPO_TOKEN` with the
+value from Coveralls.
+
 ## Ansible Requirement File Generation Tool 🧻🛠 ##
```
You are breaking the rule of using sentence case for headers.

hillaryj replied:
Fixed in https://github.com/cisagov/development-guide/pull/44/commits/f4bbeece384fc59b14ab45a60999c2417ea0d1b1 as part of the old header mop-up.

jsf9k commented on project_setup/branch-protection.md:
```diff
@@ -0,0 +1,36 @@
+    - Required approving reviews: `2`
+    - [ ] Dismiss stale pull request approvals when new commits are pushed
+    - [X] Require review from Code Owners
+    - [ ] Restrict who can dismiss pull request reviews
```
Historically we have always checked this box, I believe.

hillaryj replied:
This was based on an existing repo - while I haven't done an exhaustive survey, both `cool-system` and `development-guide` don't have it checked, while `ansible-role-cobalt-strike` and `teamserver-packer` do have it checked.

I'm in favor of standardizing on whatever the team prefers.

dav3r replied:
I prefer that we do restrict who can dismiss PR reviews.

jsf9k replied:
I agree with @dav3r that we should restrict who can dismiss PR reviews.  Checking this box gets rid of the possibility that someone simply dismisses a request for changes that he or she disagrees with instead of hashing out the issue with the reviewer in the PR.

hillaryj replied:
Resolved in https://github.com/cisagov/development-guide/pull/44/commits/18a952c15253fae65b5c657a3bea687c56f4fdcb

jsf9k commented on project_setup/branch-protection.md:
```diff
@@ -0,0 +1,36 @@
+    - [ ] Restrict who can dismiss pull request reviews
+  - [X] Require status checks to pass before merging
+  - [X] Require branches to be up to date before merging
+    - TBD note about status checks
```
Reminder that this needs to be fleshed out.

hillaryj replied:
How often do we un-check a status check as required? In general, they should all be checked/required, afaik, but do we have any common exceptions that y'all know of?

dav3r replied:
It's pretty rare that we uncheck a status as required.  These days, I think it only happens on older repos that have older-style checks still enabled/required (i.e. ones that we have made optional on newer repos like LGTM or Coveralls).  Another time we might disable a check is if the service that performs the check is down or otherwise malfunctioning.

jsf9k replied:
It's pretty rare, as @dav3r said.  There are some older repos that have some erroneous LGTM javascript checks that always fail, and we uncheck those.  (They can only be turned off by sending an email to LGTM and asking them to manually sort it out.)  New repos should not have this issue.  

On very rare occasions we manually disable checks to merge a PR in order to fix something critical that is broken in production, if the checks are failing due to the GitHub Actions service, say, being down.

Both cases are the exception rather than the rule, though, so I don't think they need to be mentioned.

jsf9k replied:
An item that probably _should_ be mentioned is that the list of status checks doesn't fully populate until a PR has been created.  This is a minor problem when one creates the repo by pushing directly to `develop` instead of pushing to another branch - the checkboxes in that list can't be checked until the first PR is created.

This problem could be alleviated if we did this (which I only very rarely do myself right now):
1. Create the repo from a skeleton via our `skeleton` tool
2. Push the repo to `develop` _before making any changes from the skeleton code_
3. Create a new branch called `first_commits` or something like that and customize the skeleton, add code, etc. until the new repo is ready
4. Create a PR to merge `first commits` into `develop`

Another advantage of this approach is that it is easy to see in that first PR where the repo creator modified the skeleton code when setting up the new repo.

I don't think it is much extra work to start doing things this way, and if we document the process and scold each other it will soon become habit.  Thoughts, @cisagov/team-ois?

dav3r replied:
I like @jsf9k's solution a lot- it gets a 👍 from me.  We might be able to update our [`skeleton` tool](https://github.com/cisagov/development-guide/blob/develop/project_setup/scripts/skeleton) to make this process even simpler.

hillaryj replied:
I like this and am documenting it now - do we want to say to merge the initial PR without reviews? That would mirror how we've just been pushing to `develop`.

dav3r replied:
I think the initial PR should still be reviewed- it's nice for visibility to the rest of the team, plus good practice in general in case there's something wonky in there.

jsf9k replied:
I agree with @dav3r 1000%.  I don't like the way we just push unreviewed code to `develop` and run with it when starting a new repo.

This would also be a good habit to get into as we hire more folks and make more use of contracting resources.  We don't want randos we hardly know and don't trust creating repos and pushing to `develop` without any review.

hillaryj replied:
Added in https://github.com/cisagov/development-guide/pull/44/commits/a645ed5182a376e39efe9cd9e3b4f0c96f33b4c7

jsf9k commented on project_setup/README.md:
```diff
@@ -1,5 +1,15 @@
+Once you've set up a repository, make sure to enable
+[branch protection](branch-protection.md).
+
 ## Skeleton Tool 💀🛠 ##
```
Sentence case for headers.

hillaryj replied:
Thanks - I'm going back and checking the existing files and headers for sentence case compliance.

hillaryj replied:
Fixed in https://github.com/cisagov/development-guide/pull/44/commits/f4bbeece384fc59b14ab45a60999c2417ea0d1b1, gotta catch em all!

jsf9k commented on project_setup/README.md:
```diff
@@ -1,5 +1,15 @@
 # Project Setup #
 
+We recommend you use a skeleton for all new repositories following the
+directions below.
```
Less awkward:
```suggestion
We recommend you follow the directions below and use a skeleton for
all new repositories.
```

jsf9k commented on project_setup/README.md:
```diff
@@ -1,5 +1,15 @@
+We recommend you use a skeleton for all new repositories following the
+directions below.
+
+For repositories from skeletons, run `pre-commit install` to run linting
```
```suggestion
For repositories created from skeletons, run `pre-commit install` to run linting
```

jsf9k reviewed (commented):
This is an excellent start.  Thank you for taking this on.  I'll happily re-review as this gets closer to merging.  (These textual PRs tend to go through a lot of revision.)

I made some suggestions.  This style guide goes hand in hand with the work @mcdonnnj has been doing to incorporate out Markdown style into our Markdown linter(s).

hillaryj committed:
Apply suggestions from code review

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

hillaryj committed:
Convert existing headers to sentence case

Follow our own advice!

hillaryj committed:
Clarify heading guidance

hillaryj committed:
Update branch-protection.md

Restrict who can dismiss PR reviews and flesh out status checks note

hillaryj committed:
Direct new repos to use an initial PR

This update is in response to discussion here: https://github.com/cisagov/development-guide/pull/44#discussion_r558366304

We want to direct users to make an initial branch and PR with their code instead of directly committing to `develop` before enabling branch protection.

Also adds a table of contents to the project setup guide and clarifies a point about status checks in the branch protection guide.

hillaryj committed:
Clarify capitalization including technical protocol abbreviations

hillaryj commented:
I added a section on capitalization of abbreviations such as HTTPS, from discussion in https://github.com/cisagov/cool-assessment-terraform/pull/93.

dav3r commented on config/gitleaks-guide.md:
```diff
@@ -1,4 +1,4 @@
-# Using gitleaks to check for PII and credentials #
+# Check for PII and Credentials with `gitleaks` #
```
```suggestion
# Check for PII and credentials with `gitleaks` #
```

I think this is what we are doing, right?  Also, do we want to expand "PII" (either in the header or in the text below) for those who are unfamiliar with the acronym?

jsf9k reviewed (approved):
Strong work, @hillaryj!  Thank you for carrying this PR over the finish line!

dav3r commented on project_setup/README.md:
```diff
@@ -87,20 +128,71 @@ Use the following commands to push the new repository to github:
-Once the `skeleton` tool is complete, inspect the new repository for accuracy.
+## Create and publish the GitHub repository ##
+
+Once the `skeleton` tool has run, the next step is to publish GitHub remote repository.
```
```suggestion
Once the `skeleton` tool has run, the next step is to publish to a GitHub remote repository.
```

dav3r commented on project_setup/README.md:
````diff
@@ -87,20 +128,71 @@ Use the following commands to push the new repository to github:
+```
+
+Edit the existing files and add your own as needed, then commit your changes,
+push them, and create a PR via GitHub or the command-line for your teammates
+to review.
+
+GitHub only populates its list of status checks once a pull request (PR)
+has been created so checks can run against changes. Status checks are
````
We should expand "PR" where we first use it, not the second place we use it.

```suggestion
push them, and create a pull request (PR) via GitHub or the command-line for
your teammates to review.

GitHub only populates its list of status checks once a PR
has been created so checks can run against changes. Status checks are
```

hillaryj replied:
Good catch!

dav3r reviewed (approved):
Awesome work on this @hillaryj - thanks for documenting it all! 🦖 

I did note a few minor suggestions that I'd like to see considered before merging.

hillaryj committed:
Apply suggestions from code review

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

hillaryj committed:
Fix linting error from suggestion

hillaryj committed:
Explain PII abbreviation
